src/Microsoft.OpenApi/Validations/Visitors/VisitorBase.cs

```diff
@@ -0,0 +1,52 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license. 
+
+using System.Diagnostics;
+using System.Linq;
+using Microsoft.OpenApi.Interfaces;
+
+namespace Microsoft.OpenApi.Validations.Visitors
+{
+    /// <summary>
+    /// The base visitor class for Open Api elements.
+    /// </summary>
+    internal abstract class VisitorBase<T> : IVisitor where T : IOpenApiElement
+    {
+        /// <summary>
+        /// Visit the element itself.
+        /// </summary>
+        /// <param name="context">The validation context.</param>
+        /// <param name="item">The element.</param>
+        public void Visit(ValidationContext context, object item)
+        {
+            Debug.Assert(item is T, "item should be " + typeof(T));
+
+            var rules = context.RuleSet.Where(r => r.ElementType == typeof(T));
+            foreach (var rule in rules)
+            {
+                rule.Evaluate(context, item);
+            }
+
+            T typedItem = (T)item;
+            this.Next(context, typedItem);
+        }
+
+        /// <summary>
+        /// Visit the children.
+        /// </summary>
+        /// <param name="context">The validation context.</param>
+        /// <param name="element">The element.</param>
+        protected virtual void Next(ValidationContext context, T element)
+        {
+            IOpenApiExtensible extensbile = element as IOpenApiExtensible;
+            if (extensbile != null)
+            {
+                var rules = context.RuleSet.Where(r => r.ElementType == typeof(IOpenApiExtensible));
+                foreach (var rule in rules)
+                {
+                    rule.Evaluate(context, extensbile);
+                }
+            }
```

Comment:
Why is this not done in Visit above?

Reply:
:) My design is 

1. Visit() method is used to verify the `itself` properities
2. Next() method is used to verify the `Children` element.

So, I think Extension belongs to the `Children` scope. What do you think?

Reply:
:) My design is 
1. Visit() method is used to verify the itself properities
2. Next() method is used to verify the Children element.
So, I think Extension belongs to the Children scope. What do you think?

---
In reply to: [157650373](https://github.com/Microsoft/OpenAPI.NET/pull/152#discussion_r157650373) [](ancestors = 157650373)

Reply:
Why does Extension fall into the Children scope? Isn't it one of the properties?

---
In reply to: [157838009](https://github.com/Microsoft/OpenAPI.NET/pull/152#discussion_r157838009) [](ancestors = 157838009,157650373)